text/0000-item-level-scopes.md

```diff
@@ -0,0 +1,181 @@
+this:
+
+```rust
+#[cfg(debug_assertions)]
+use a;
+#[cfg(debug_assertions)]
+use b;
+#[cfg(debug_assertions)]
+use c;
+#[cfg(debug_assertions)]
+use d
+```
+
+The above example is quite verbose because the `#[cfg...]` line must be
+repeated for each `use` line. This RFC proposes that the following should be
+possible:
+
+```rust
+#[cfg(debug_assertions)] {
+    use a;
+    use b;
+    use c;
+    use d;
+}
+```
+
+The latter is more concise and thus easier to read and write.
+
+# Guide-level explanation
+[guide-level-explanation]: #guide-level-explanation
+
+An Item-level scope is allowed wherever an Item may appear. Item-level scopes
+are useful for (e.g.) compile-time guarding a collection of `use` statements.
+
+The following:
+
+```rust
+#[cfg(debug_assertions)]
+use a;
+#[cfg(debug_assertions)]
+use b;
+#[cfg(debug_assertions)]
+use c;
+#[cfg(debug_assertions)]
+use d
+```
+
+Can be written more concisely as:
+
+```rust
+#[cfg(debug_assertions)] {
+    use a;
+    use b;
+    use c;
+    use d;
+}
+```
+
+Item-level scopes may be nested, however note that only Items can appear
+inside. Other elements, such as control-flow expressions, may not appear inside
+Item-level scopes.
+
+# Reference-level explanation
+[reference-level-explanation]: #reference-level-explanation
+
+Implementation amounts to implementing a new kind of scope, which is itself an
+Item.
+
+If a non-Item appears in a Item-level scope, then an appropriate error
+message should be given. For example the Rust file:
+
+```rust
+#[cfg(debug_assertions)] {
+    use a;
+    for _ in 0..3 {
+        println!("x");
+    }
+}
+```
+
+Should fail with an error message like:
+
+```
+error: expected item, found `for`
+ --> src/main.rs:3:5
+  |
+3 |     for _ in 0..3 {
+  |     ^^
+  |
+  = note: Only items can appear in item-level scopes.
+
+```
+
+It should also be possible to nest Item-level scopes. For example, the
+following should be valid:
+
+```rust
+#[cfg(debug_assertions)] {
+    use a;
+    use b;
+    #[cfg(target_os = "linux")] {
+        use c;
+        use d;
+    }
+}
+```
+
+# Drawbacks
+[drawbacks]: #drawbacks
+
+Because an Item-level scope looks the same as a regular scope (i.e. `{ ... }`)
+users may think that they can insert arbitrary expressions. A good error
+message should dispel this concern.
+
+# Rationale and alternatives
+[alternatives]: #alternatives
+
+The [cfg-if](https://crates.io/crates/cfg-if) crate allows the following:
+
+```rust
+cfg_if! {
+    if #[cfg(debug_assertions)] {
+        use a;
+        use b;
+        use c;
+        use d;
+    }
+}
+```
+
+However, this is more verbose, and requires using an external crate.
+
+The closest you can get without using a crate is by using a submodule trick:
+
+```rust
+#[cfg(debug_assertions)]
+use self::imports::*;
+#[cfg(debug_assertions)]
+mod imports {
+    pub(super) use a;
+    pub(super) use b;
+    pub(super) use c;
+    pub(super) use d;
+}
+```
+
+Which seems overly verbose and complicated.
```

Comment:
A better example of a currently usable alternative is:

```rust
#[cfg(debug_assertions)]
use ::{a, b, c, d};
```

which I think is definitely good enough for this specific use case.

Reply:
Strongly agreed, this is what I was trying to say with my previous comment linking to https://github.com/rust-lang/rfcs/pull/2128. If the only motivation for "item-level scopes" is to apply an attribute to several `use`s, then your syntax as enabled by that RFC is the better solution.

Reply:
It depends if people would like to (for example), batch a load of function defs and/or data structures into a guarded scope. Do you think that might be useful? If not, then perhaps this RFC is superfluous?

Reply:
I can imagine one such scenario: for platform abstractions, where each platform might have it's own data structures and functions.

```
#[cfg(target_platform = "linux")] {
    use something::only::linux::uses;
    use more::linux::stuff;

    const SOME_SPECIFIC_PATH: &'static str = "/sys/blah/flibble/linux_specific123";

    struct LinuxThingy1 {
       ...
    }

    impl LinuxThingy1 {
         ...
    }

    struct LinuxThingy2 {
        ...
    }

    impl LinuxThingy2 {
         ...
    }
}

#[cfg(target_platform = "windows")] {
    ... lots of completely different imports and data structures...
}

#[cfg(target_platform = "freebsd")] {
    ... lots of yet different imports and data structures...
}
```

Reply:
That's a much better example, another relevant example could be replacing the [`if_std!`](https://github.com/rust-lang-nursery/futures-rs/blob/53796cc27ead6988478fa09681750386fb3998a2/futures-core/src/lib.rs#L16-L21) macro [used in `futures`](https://github.com/rust-lang-nursery/futures-rs/blob/53796cc27ead6988478fa09681750386fb3998a2/futures-core/src/stream/mod.rs#L73-L101).